### PR TITLE
Restrict RBAC for Operator

### LINKED ISSUE #106: Principle of least privilege should be applied in operator (Cluster)Roles
**Describe the bug**
Today the operator is by default granted all permissions on all resources clusterwide by a [blanket ClusterRole](https://github.com/knative/serving-operator/blob/master/config/role.yaml#L60)

**Expected behavior**
Only specifically those permissions which are necessary in order to create a functional Knative serving setup should be added, using [escalate and/or bind](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#privilege-escalation-prevention-and-bootstrapping) to create roles or rolebindings that the operator itself does not need explicitly, but which are needed transitively by knative serving.

**To Reproduce**
1. Install Knative Eventing Operator.
1. kubectl describe clusterrole knative-eventing-operator
```
Name:         knative-eventing-operator
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"rbac.authorization.k8s.io/v1","kind":"ClusterRole","metadata":{"annotations":{},"name":"knative-eventing-operator"},"rules":...
PolicyRule:
  Resources  Non-Resource URLs  Resource Names  Verbs
  ---------  -----------------  --------------  -----
  *.*        []                 []              [*]
```

Sibling bug of https://github.com/knative/serving-operator/issues/282

Going to use this to attempt to solve both in a way that generalizes well.

### PR DESCRIPTION
**Pull request by Cynocracy**
Fixes #106

#150 is a base of this, you can solely review the YAML here
(and not the reconciler change) though you're welcome to review
that as well.

## Proposed Changes

By using the `escalate` verb on clusterroles, we allow the Operator
to have the ability to create any clusterrole. This prevents us from
needing to grant it `cluster-admin`.

Additionally, by having the Operator create an aggregated clusterrole
for itself (aggregating all clusteroles created by Knative Eventing),
we pick up any permissions that the upstream dictate as necessary
by any component.

Any roles that are not created by upstream, but which are bound as
part of install are added with the `bind` verb to allow the Operator
to bind them without necessarily having the permissions they grant.

Some other permissions are needed to support controller infra
and are explicitly included, as well as some specific to the Operator
itself. An attempt at grouping these by sections is included, though
may be slightly out of sync (in particular: Controller infra / Specific
to this Operator may overlap).

Overall, this should keep the Operator roughly operating only on
Knative Eventing resources (or things which can be operated
on by upstream).

**Release Note**

```release-note
Change RBAC to minimize permissions of Operator service account.
```

### CODE REVIEW

**Comment by Cynocracy** [comment]
This should begin to pass once we regen the eventing YAML

**Comment by Cynocracy** [comment]
{"level":"info","ts":"2020-03-06T23:29:47.135Z","logger":"eventing_operator.knativeeventing-controller.event-broadcaster","caller":"record/event.go:274","msg":"Event(v1.ObjectReference{Kind:\"KnativeEventing\", Namespace:\"operator-tests\", Name:\"knative-eventing\", UID:\"cdb13e43-2533-4246-8050-3886920854ef\", APIVersion:\"operator.knative.dev/v1alpha1\", ResourceVersion:\"2475\", FieldPath:\"\"}): type: 'Warning' reason: 'InternalError' clusterroles.rbac.authorization.k8s.io \"source-observer\" not found","commit":"025ff25","knative.dev/controller":"knativeeventing-controller"}
{"level":"info","ts":"2020-03-06T23:29:47.135Z","logger":"eventing_operator.knativeeventing-controller","caller":"controller/controller.go:363","msg":"Reconcile failed. Time taken: 3.20852151s.","commit":"025ff25","knative.dev/controller":"knativeeventing-controller","knative.dev/traceid":"26604652-ebba-4703-9fa8-a46d3c00f35a","knative.dev/key":"operator-tests/knative-eventing"}
{"level":"info","ts":"2020-03-06T23:30:28.095Z","logger":"eventing_operator.knativeeventing-controller","caller":"knativeeventing/knativeeventing.go:112","msg":"Reconciling KnativeEventing","commit":"025ff25","knative.dev/controller":"knativeeventing-controller","Request.Namespace":"operator-tests","Request.Name":"knative-eventing","status":{"conditions":[{"type":"InstallSucceeded","status":"Unknown","lastTransitionTime":"2020-03-06T23:28:29Z"},{"type":"Ready","status":"False","lastTransitionTime":"2020-03-06T23:28:36Z","reason":"Manifest Installation","message":"clusterroles.rbac.authorization.k8s.io \"source-observer\" not found"}]}}
{"level":"error","ts":"2020-03-06T23:30:31.326Z","logger":"eventing_operator.knativeeventing-controller","caller":"controller/controller.go:376","msg":"Reconcile error","commit":"025ff25","knative.dev/controller":"knativeeventing-controller","error":"clusterroles.rbac.authorization.k8s.io \"source-observer\" not found","stacktrace":"knative.dev/eventing-operator/vendor/knative.dev/pkg/controller.(*Impl).handleErr\n\tknative.dev/eventing-operator/vendor/knative.dev/pkg/controller/controller.go:376\nknative.dev/eventing-operator/vendor/knative.dev/pkg/controller.(*Impl).processNextWorkItem\n\tknative.dev/eventing-operator/vendor/knative.dev/pkg/controller/controller.go:362\nknative.dev/eventing-operator/vendor/knative.dev/pkg/controller.(*Impl).Run.func2\n\tknative.dev/eventing-operator/vendor/knative.dev/pkg/controller/controller.go:310"}

**Comment by Cynocracy** [comment]
Whoops, big mistake in the existing role. This needs an update and retest.

**Comment by Cynocracy** [comment]
New issue seems to be the use of the operator-tests namespace, iirc this was dropped for kserving.

**Comment by Cynocracy** [comment]
Bah humbug, let's see what the problem is.

**Comment by Cynocracy** [comment]
msg: "Event(v1.ObjectReference{Kind:"KnativeEventing", Namespace:"operator-tests", Name:"knative-eventing", UID:"41eff4b4-2aee-478d-99dc-9baf59462f1d", APIVersion:"operator.knative.dev/v1alpha1", ResourceVersion:"2200", FieldPath:""}): type: 'Warning' reason: 'InternalError' serviceaccounts "eventing-controller" is forbidden: User "system:serviceaccount:default:knative-eventing-operator" cannot get resource "serviceaccounts" in API group "" in the namespace "operator-tests"" 

Aha, right, the namespace thing, maybe I can test it out on a different namespace

**Comment by Cynocracy** [comment]
Very odd that both tests pass locally.

I'm going to tilt at that windmill for a bit.

**Comment by Cynocracy** [comment]
Oh :) That was a silly reason.

**Comment by Cynocracy** [comment]
/assign @n3wscott

Assuming my last minute diff-tweaking didn't break this, this should be ready to review :)

The role defined in role.yaml is a strict subset of that in serving-operator, where in that repo
we have a few 'bind's on roles which are used but not created by the operator (now required)
and a 'cleanup' section where serving operator needs permission to delete older resources it now
routinely cleans up.
In the future, if we move to a job based cleanup, those can be defined as one-offs.

If the Operator repos and/or controllers are to merge, this dramatically simplifies the collapse of 
the roles needed.

File: config/role.yaml [hunk 1]
```diff
@@ -1,19 +1,91 @@
...
+kind: ClusterRole
+metadata:
+  name: knative-eventing-operator-aggregated
+aggregationRule:
+  clusterRoleSelectors:
+# This (along with escalate below) allows the Operator to pick up any
+# roles that are provided to the admin of the cluster by knative eventing
+# automatically.
+  - matchExpressions:
+      - {key: eventing.knative.dev/release, operator: Exists}
```
Comment (n3wscott) [inline config/role.yaml]: that is pretty clever.

Comment (Cynocracy) [reply config/role.yaml]: thanks :)

**Pull request merged**